src/pointer-analysis/value_set_dereference.cpp

```diff
@@ -201,13 +201,25 @@ bool value_set_dereferencet::dereference_type_compare(
   const typet &object_type,
   const typet &dereference_type) const
 {
-  if(dereference_type.id()==ID_empty)
-    return true; // always ok (anything can be dereferenced to void type)
-
+  // check if the two types have matching number of ID_pointer levels, with
+  // the dereference type eventually pointing to void; then this is ok
+  // for example:
+  // - dereference_type=void is ok (no matter what object_type is);
+  // - object_type=(int *), dereference_type=(void *) is ok;
+  // - object_type=(int **), dereference_type=(void **) is ok;
+  // - object_type=(int *), dereference_type=(void **) is not ok;
+  typet object_unwrapped = object_type;
+  typet dereference_unwrapped = dereference_type;
```

Comment:
I would have used pointers here, but I won't insist. There are a couple of places in the codebase that dig through array or pointer types, and those tend to use pointers to avoid repeated construction of a non-trivial object.

Reply:
I tend to insist that pointers should be used here.